File: src/main/kotlin/broker/Broker.kt
package broker

import akka.actor.ActorSystem
import akka.actor.PoisonPill
import akka.http.javadsl.ConnectHttp
import akka.http.javadsl.Http
import akka.http.javadsl.model.HttpRequest
import akka.http.javadsl.model.HttpResponse
import akka.http.javadsl.model.StatusCodes
import akka.http.javadsl.model.ws.Message
import akka.http.javadsl.model.ws.UpgradeToWebSocket
import akka.stream.ActorMaterializer
import akka.stream.OverflowStrategy
import akka.stream.javadsl.Sink
import akka.stream.javadsl.Source

fun main() {

    val system = ActorSystem.create("quicks-broker")

    try {

        val mat = ActorMaterializer.create(system)

        val connectionsManager = system.actorOf(ConnectionsManager.props(), "c-manager")

        val serverBindingFuture = Http.get(system)
            .bindAndHandleSync({ request ->

                if (request.uri.path() != "/ws") {
                    request.discardEntityBytes(mat)
                    return@bindAndHandleSync HttpResponse.create()
                        .withStatus(StatusCodes.NOT_FOUND)
                        .withEntity("Unknown resource !")
                }

                val header = request.wsUpgradeHeader() ?:
                return@bindAndHandleSync HttpResponse.create()
                    .withStatus(StatusCodes.BAD_REQUEST)
                    .withEntity("Ws upgrade header not found !")

                // getting the actor responsible for sending outgoing messages
                val source = Source
                    .actorRef<Message>(10, OverflowStrategy.fail())
                    .preMaterialize(mat)

                // passing the source actor to the connection actor to enable sending messages from it
                val connectionActor = system.actorOf(ConnectionActor.props(source.first(), connectionsManager))

                // routing incoming messages to the connection actor
                val sink = Sink.actorRef<Message>(connectionActor, PoisonPill.getInstance())

                return@bindAndHandleSync header.handleMessagesWith(sink, source.second())

        }, ConnectHttp.toHost("127.0.0.1", 8080), mat)

        println("Press ENTER to exit the system")
        readLine()

        serverBindingFuture
            .thenCompose { it.unbind() }
            .thenAccept { system.terminate() }

    } finally {
        system.terminate()
    }
}

fun HttpRequest.wsUpgradeHeader(): UpgradeToWebSocket? {
    return headers.firstOrNull { it is UpgradeToWebSocket } as? UpgradeToWebSocket
}

File: src/main/scala/com/quicks/broker/Broker.scala
package com.quicks.broker

import java.util.Base64

import akka.NotUsed
import akka.actor.{ActorSystem, Props}
import akka.http.scaladsl.Http
import akka.http.scaladsl.model.headers.RawHeader
import akka.http.scaladsl.model.ws.Message
import akka.http.scaladsl.model.{ContentTypes, HttpMethods, HttpRequest}
import akka.http.scaladsl.server.Directives._
import akka.http.scaladsl.settings.ServerSettings
import akka.stream.scaladsl.{Flow, Sink, Source}
import akka.stream.{ActorMaterializer, OverflowStrategy}
import com.quicks.broker.ConnectionAgent._

import scala.concurrent.Await
import scala.concurrent.duration._
import scala.io.StdIn

object Broker {

  def main(args: Array[String]) {

    implicit val system = ActorSystem()
    implicit val materializer = ActorMaterializer()
    implicit val executionContext = system.dispatcher
    val serverSettings = ServerSettings(system)

    val customWebsocketSettings =
      serverSettings.websocketSettings
      .withPeriodicKeepAliveMaxIdle(1.second)

    val customSettings =
      serverSettings.withWebsocketSettings(customWebsocketSettings)

    val connManager = system.actorOf(Props[ConnectionsManager])

    def newUser(): Flow[Message, Message, NotUsed] = {

      val connAgent = system.actorOf(Props(new ConnectionAgent(connManager)))

      val in: Sink[Message, NotUsed] = Sink.actorRefWithAck(
        connAgent,
        Init,
        Ack,
        Complete,
        it => Error(it)
      )

      val out = Source.actorRef(10, OverflowStrategy.fail)
        .mapMaterializedValue( sender => {
          connAgent ! SenderReceived(sender)
          NotUsed
        })

      Flow.fromSinkAndSource(in, out)
    }

    val route =
      get {
        path("ws") {
          handleWebSocketMessages(newUser())
        }
      } ~ get {
        path("ice") {
          val request = HttpRequest(uri = "https://global.xirsys.net/_turn/Quicks")
            .withMethod(HttpMethods.PUT)
            .addHeader(RawHeader(
              "Authorization",
              "Basic " + Base64.getEncoder.encodeToString(sys.env("XIRSYS_TOKEN").getBytes()))
            )
            .withEntity(ContentTypes.`application/json`, "{\"format\": \"urls\"}")
          onSuccess(Http().singleRequest(request)) { it =>
            complete(it)
          }
        }
      }

    val binding = Await.result(Http()
      .bindAndHandle(route, "127.0.0.1", 8080, settings = customSettings), 3.seconds)

    // the rest of the sample code will go here
    println("Started server at 127.0.0.1:8080, press enter to kill server")
    StdIn.readLine()
    system.terminate()

  }

}
